crates/wasmtime/src/runtime/vm/interpreter.rs

```diff
@@ -12,17 +12,45 @@ use wasmtime_environ::{BuiltinFunctionIndex, HostCall, Trap};
 /// Interpreter state stored within a `Store<T>`.
 #[repr(transparent)]
 pub struct Interpreter {
-    /// Pulley VM state, stored behind a `Box<T>` to make the storage in
-    /// `Store<T>` only pointer-sized (that way if you enable pulley but don't
-    /// use it it's low-overhead).
-    pulley: Box<Vm>,
+    /// Pulley VM state, stored in a `StoreBox<T>`.
+    ///
+    /// This representation has a dual purpose of (a) having a low overhead if
+    /// pulley is disabled (just a null pointer) and (b) enabling safe access to
+    /// the `Vm` in the face of recursive calls.
+    ///
+    /// For (b) that's the most tricky part of this, but the basic problem looks
+    /// like:
+    ///
+    /// * The host initially executes some WebAssembly.
+    /// * This acquires a `&mut Vm` and does some execution.
+    /// * The WebAssembly then invokes the host.
+    /// * This bottoms out in `CallIndirectHost` which means that we'll do a
+    ///   dynamic dispatch to a function pointer in pulley registers.
+    /// * The function we call gets unfettered access to `StoreContextMut<T>`
+    /// * When the function returns our original `&mut Vm` pointer is
+    ///   invalidated, so it has to be re-acquired.
+    ///
+    /// The usage of `StoreBox` here solves this conundrum by storing the
+    /// `InterpreterRef` at-rest as a `NonNull<Vm>` as opposed to a `&mut Vm`.
+    /// This is required to model how after a host call the VM state must be
+    /// re-acquire from store state to re-assert that it has an exclusive
+    /// borrow.
+    ///
+    /// This in turn models how VM state could be modified as part of the
+    /// recursive function call, for example with another VM execution itself.
+    ///
+    /// Note that the safety of this all relies not only on correctly managing
+    /// this pointer but it also requires that this pointer is never
+    /// deallocated. This private field is never overwritten which ensures such
+    /// a guarantee.
```

Comment:
Never deallocated *for the lifetime of the store*, right? We aren't leaking Pulley VMs indefinitely. Would be good to clarify that in this comment.